src/main/scala/io/iohk/ethereum/vm/VM.scala

```diff
@@ -1,36 +1,88 @@
 package io.iohk.ethereum.vm
 
+import akka.util.ByteString
+import io.iohk.ethereum.domain.Address
 import io.iohk.ethereum.utils.Logger
 
 import scala.annotation.tailrec
 
-/**
-  * Entry point to executing a program.
-  */
-class VM extends Logger {
+class VM[W <: WorldStateProxy[W, S], S <: Storage[S]] extends Logger {
+
+  type PC = ProgramContext[W, S]
+  type PR = ProgramResult[W, S]
+  type PS = ProgramState[W, S]
 
   /**
-    * Executes a program
+    * Executes a top-level program (transaction)
     * @param context context to be executed
     * @return result of the execution
    */
-  def run[W <: WorldStateProxy[W, S], S <: Storage[S]](context: ProgramContext[W, S]): ProgramResult[W, S] = {
-    PrecompiledContracts.runOptionally(context).getOrElse {
-      val finalState = run(ProgramState(context))
-      ProgramResult[W, S](
-        finalState.returnData,
-        finalState.gas,
-        finalState.world,
-        finalState.addressesToDelete,
-        finalState.logs,
-        finalState.internalTxs,
-        finalState.gasRefund,
-        finalState.error)
+  def run(context: ProgramContext[W, S]): ProgramResult[W, S] = {
+    {
```

Comment:
still needed?

Reply:
Well, doesn't hurt. It's on `trace`

Reply:
I meant just the curly brace

Reply:
Ahh, I added the curly brace because of the `import context._` to avoid polluting the remaining scope. Especially now with the match statement (`case Some(recipientAddr)`) there would be shadowing, which is something I try to avoid.